goodtablesio/integrations/s3/blueprint.py

```diff
@@ -170,3 +95,130 @@ def create_job():
     _run_validation(bucket, job_id)
 
     return jsonify({'job_id': job_id})
+
+
+# API
+
+# TODO:
+# it should be synced with general
+# approach we use for API (see api blueprint)
+
+@s3.route('/api/bucket')
+@login_required
+def api_bucket_list():
+    error = None
+    buckets = get_user_buckets(current_user.id)
+    buckets = [bucket.to_api() for bucket in buckets]
+    return jsonify({
+        'buckets': buckets,
+        'error': error,
+    })
+
+
+@s3.route('/api/bucket/<bucket_name>')
+@login_required
+def api_bucket(bucket_name):
+    try:
+        code = 200
+        buckets = get_user_buckets(current_user.id)
+        bucket = [item for item in buckets if item.name == bucket_name][0].to_api()
+        error = None
+    except Exception:
+        code = 404
+        bucket = None
+        error = 'Not Found'
+    return (jsonify({
+        'bucket': bucket,
+        'error': error,
+    }), code)
+
+
+@s3.route('/api/bucket', methods=['POST'])
+@login_required
+def api_bucket_add():
+    error = None
+    payload = request.get_json()
+    access_key_id = payload.get('access-key-id')
+    secret_access_key = payload.get('secret-access-key')
+    bucket_name = payload.get('bucket-name')
+
+    # Check input fields
+    if not access_key_id or not secret_access_key or not bucket_name:
+        error = 'Missing fields'
+
+    # Get bucket
+    if not error:
+        source = database['session'].query(S3Bucket).filter(
+            S3Bucket.name == bucket_name).one_or_none()
+        if source and source.active:
+            error = 'Bucket already exists'
+
+    # Setup bucket on aws
+    if not error:
+        try:
+            # TODO: catch all errors inside util function
+            success, message = set_up_bucket_on_aws(
+                access_key_id, secret_access_key, bucket_name)
+            if not success:
+                error = 'Error setting up bucket integration. {0}'.format(message)
+        except Exception:
```

Comment:
I was thinking about it and you're right. We anyway should catch server error in client so there is no reason to *guarantee* (anyway not possible because of other failure reasons) api always returning json.